Add Window.Owner. Patch for Issue #319

Panzerfather opened the pull request:
Setup parent when using Dialog.ShowModal

cwensley commented:
Thanks for the pull request!  It is really appreciated.  A few questions:

What happens if you pass a Button instance to ShowModal?  The parent will still be null.  

Also, using the existing Parent property is probably not a good idea, as it is used to tell which control is the containing parent in the UI hierarchy, which is a different purpose for this.. It might be better to have an Owner property (of type Window) added to the Dialog class that would be set to the control.ParentWindow when ShowModal is called.

Thoughts?

Panzerfather commented:
I'm not quite sure what you mean with "what happens if you pass a Button instance to ShowModal?". Eto.Button doesn't inherit a ShowModal method.

The former patch was using the as-cast, so if the parent is null, it will be set to null. No error will occur. It doesn't matter which control inherits the method.

But I agree with your idea of integrating an owner property, so I've done this in my latest patch. The patches for Android, IOS and WinRT are untested, but should be functional.

cwensley commented:
@Panzerfather, thanks for the feedback!  I'll try to explain what I am thinking regarding those..

The problem with making the Owner property writable is that it is not doing anything when you set it.  Say for example you create a Dialog, set it's Owner, then call ShowModal(), it won't appropriately set the owner for each platform, whereas as a consumer of the API I would expect it to.

If we want to make it writable, then the handler implementations should use the property when showing the dialog instead of using the Control parameter (e.g. the parent parameter should then be removed from Dialog.IHandler.ShowModal and ShowModalAsync).

As for moving this to Dialog - Wpf and WinForms don't have a separation of Dialog (modal) vs. Form (modeless) - this property won't do anything on a Form, so it should probably not be available there.  You could still do as you describe by closing all non-modal windows by closing all Form windows and skipping Dialog windows.

Panzerfather commented:
@cwensley, thanks for the explanation. I removed the Handler-Code in the latest commit and added a property and an event to the Window class. I also made small changes to the Dialog and Form classes

Let me show and explain why it makes sence to use the Window class and not only the Dialog class for implementation:

![application](https://cloud.githubusercontent.com/assets/11777736/7037181/82cfc364-dd9e-11e4-9e38-f7eaee47f012.png)

As you can see in the screenshot there is a Main-Form which you should imagine as a Control-Window. From this form you can run and show with multiple windows as usual.
But the main difference is that if the Owner-Property is setup in the Window-Class the Sub-[Window/-Dialog/Form] can interact with the Main-Form without having to maintain e.g. your own list of opened windows.
e.g. the "Close all windows which have me as owner"-Button have just following code:
```C#
protected void OnButtonCloseAllClick(object sender, EventArgs e)
{
	// Have to be list or equivalent to avoid enumerable changed error
	List<Window> lstWindows = Application.Instance.Windows.Where(w => w.Owner == this).ToList();

	foreach (Window winOpen in lstWindows)
	{

		winOpen.Close();

	}
}
```

And it closes all Windows which have the Main-Form as owner. All others will remain open. You could also easily iterate through a Sub-SubWindow with this approach, to close them, update them, whatever.

This example can also be done when using .NET WinForms and WPF and set the Owner-Property manually.

As a good multiplatform example, Gimp is using such a approach if you are not running it in SingleWindow-Mode. But they have to maintain there own Handling-Code, which isn't fail-safe like this one.

cwensley commented on Source/Eto/Forms/Window.cs:
```diff
@@ -118,6 +118,30 @@ protected virtual void OnLocationChanged(EventArgs e)
+		{
+			add { OwnerChangedEvent += value; }
+			remove { OwnerChangedEvent -= value; }
+		}
```
This should be changed to use the Properties.AddEvent() and Properties.RemoveEvent() so as to remove memory overhead when not handled

cwensley commented on Source/Eto/Forms/Window.cs:
```diff
@@ -118,6 +118,30 @@ protected virtual void OnLocationChanged(EventArgs e)
+		/// <param name="e">Event arguments</param>
+		protected virtual void OnOwnerChanged(EventArgs e)
+		{
+			if (OwnerChangedEvent != null)
```
Should be `Properties.TriggerEvent(OwnerChangedEvent, this, e)`

cwensley commented on Source/Eto/Forms/Form.cs:
```diff
@@ -30,8 +30,10 @@ public Form(IHandler handler)
 		/// </summary>
-		public void Show()
-		{ 
+		public void Show(Window owner = null)
```
I was thinking of reverting this in favour of using the Owner property directly.. Same with Dialog.ShowModal/Async() to deprecate the existing method(s) and add overloads without the owner parameter..  I think that provides a much cleaner api.. e.g. 
```c#
var dlg = new Dialog { Owner = this };
dlg.ShowModal();
```

Thoughts?

cwensley commented:
This looks like it needs a rebase.. the diffs show all the changes that are unrelated to the pull

Panzerfather commented:
Rebased it.

cwensley commented on Source/Eto.Gtk/Forms/GtkWindow.cs:
```diff
@@ -249,7 +249,10 @@ public override void AttachEvent(string id)
-					Control.Shown += Connector.HandleShown;
+					Control.Shown += Connector.HandleShownEvent;
+					break;
+				case Window.OwnerChangedEvent:
```
This shouldn't be necessary, this doesn't need to be a HandlerEvent..  I believe this will cause the event to be called more than once, since you are already calling it when setting the Window.Owner property..

Panzerfather replied:
If it isn't handled in the Platform-Handling, there's an exception of a missing event. That was the reason I didn't used the Properties-Event handling in f5d7b493c14c8f7b60ce2e358ec9982c8fa64dd3

Panzerfather replied:
Changed Property-Event declaration to remove platform handling code in the next commit 7f696f6f9951299fa9c838c652e94b19b502b498.

cwensley commented:
I think for the OwnerChanged event, either it needs to be handled in each of the platform handlers, or have it trigger when the property is set in Eto.Forms.Window.Owner.. it should probably not be called by both..

Panzerfather commented:
Cleaned up code and removed platform handling.

cwensley commented on Source/Eto.Gtk/Forms/GtkWindow.cs:
```diff
@@ -287,11 +287,16 @@ public void HandleDeleteEvent(object o, Gtk.DeleteEventArgs args)
 				Handler.Callback.OnShown(Handler.Widget, EventArgs.Empty);
 			}
 
+			public void HandleOwnerEvent(object sender, EventArgs e)
```
This should be removed now that we aren't using the handler event

cwensley commented on Source/Eto.Mac/Forms/MacWindow.cs:
```diff
@@ -289,6 +289,14 @@ static bool HandleWindowShouldClose(NSObject sender)
 			return !args.Cancel;
 		}
 
+		static void HandleOwnerChanged(object sender, EventArgs e)
```
Should be removed also

cwensley commented on Source/Eto.Mac/Forms/MacWindow.cs:
```diff
@@ -373,6 +384,18 @@ public override void AttachEvent(string id)
 						Control.WillMove += HandleWillMove;
 					}
 					break;
+				case Window.OwnerChangedEvent:
```
More removal.. (;

cwensley commented:
Some of the platform handling still seems to be there.. could you take a look?

Panzerfather commented:
Pushed missing cleanup, now all platform code should be removed.

cwensley commented:
Definitely getting there!   I'm still seeing a wee bit of platform code lingering.. in the Callback/ICallback, the static constructor of Window, and the Window.OwnerChangedEvent id should be private (only should be public for handler-based events).

Panzerfather commented:
This change can't be integrated. For a Callback the private keyword isn't allowed and would create compiler errors.
Also the static Window() method can't have an keyword, keyword modifiers aren't allowed for static constructors. This would also create compiler errors.

Panzerfather commented:
Made the OwnerChangedEvent constant private and added a an check to Eto.Test. Restructuring the child changing a little bit to ensure that all events will be fired. Event-Callback overloads were made later in code than the changes of the Window/Dialog itself.

cwensley commented on Source/Eto.WinRT/Forms/WpfWindow.cs:
```diff
@@ -109,6 +109,9 @@ public override void AttachEvent (string id)
 					}
 				};
 				break;
+			case Window.OwnerChangedEvent:
```
This is still some handler code that should be removed

cwensley commented on Source/Eto/Forms/Window.cs:
```diff
@@ -142,11 +165,18 @@ protected virtual void OnWindowStateChanged(EventArgs e)
 			EventLookup.Register<Window>(c => c.OnClosed(null), ClosedEvent);
 			EventLookup.Register<Window>(c => c.OnClosing(null), ClosingEvent);
 			EventLookup.Register<Window>(c => c.OnLocationChanged(null), LocationChangedEvent);
+			EventLookup.Register<Window>(c => c.OnOwnerChanged(null), OwnerChangedEvent);
```
This is only needed for handler-based events

cwensley commented on Source/Eto/Forms/Window.cs:
```diff
@@ -451,6 +502,10 @@ public new interface ICallback : Panel.ICallback
 			/// <summary>
+			/// Raises the owner changed event.
+			/// </summary>
+			void OnOwnerChanged(Window widget, EventArgs e);
```
This is also not necessary, as the handlers aren't triggering this event.

Panzerfather commented:
Made the changes, sorry for the forgotten removal of the WinRT platform code. It's a shame that VS 2013 can't handle/open WinRT-Projects on none Windows 8.x platforms! So they aren't even searchable for code-fragments. Maybe therefor I won't change code for this platform anymore. Making to much trouble when adding/removing.

cwensley commented:
No problem at all!  It happens to everyone (including myself) from time to time.  I believe Win 10 will be free upgrade for Win 7 so at least you'll have a way to do it in the future.

A squash would be really nice on this pull as well if that isn't too much trouble?

cwensley commented on Source/Eto/Forms/Window.cs:
```diff
@@ -483,6 +529,13 @@ public void OnLocationChanged(Window widget, EventArgs e)
 			/// <summary>
+			/// Raises the owner changed event.
+			/// </summary>
+			public void OnOwnerChanged(Window widget, EventArgs e)
```
One last bit to remove.. (;

Panzerfather replied:
Done it, working on a squash right now.

Panzerfather commented:
Squashed it.

cwensley commented:
Beautiful!  Nice and clean, thank you so much!  I like how you picked up on the Obsolete convention for specifying the version it was deprecated (;
